### PR TITLE
Lazily download `test262`  repository

### LINKED ISSUE #3095: Consider removing tc39/test262 submodule and download lazily
Due to how cargo handles submodules (or rather doesn't): https://github.com/rust-lang/cargo/issues/4247

a fresh cargo build with boa as git dependency `boa { git = "..."}` will download all submodules and it takes some time to download https://github.com/tc39/test262

Perhaps it's possible to download the repo lazily instead, for example:

https://github.com/dtolnay/syn/pull/183

### PR DESCRIPTION
**Pull request by HalidOdat**
This Pull Request fixes/closes #3095 .

This PR makes the downloads the `test262` repository, if we run the `boa_tester` with `run` command and we do not specify the test262 directory.

There is a con with this approach, dependabot can't suggest a updates. We could implement our own checker though.

TODO:

- [x] The commit is hard coded, add a config file for test262 default commit (maybe merge `test_ignore.toml` into it).
  - [x] Add ability to specify commit on test262 clone, via CLI commands
  - [x] Warn user that the current commit is older than test262 HEAD
- [x] Better error reporting

### CODE REVIEW

File: boa_tester/src/main.rs [hunk 1]
```diff
@@ -241,12 +241,45 @@ fn main() -> Result<()> {
...
+                    .expect("");
+
+    println!("result: {result}");
+    assert!(result.success());
+
+    println!("reset to known-good rev");
+    let result = Command::new("git")
+                    .arg("reset")
+                    .arg("--hard")
+                    .arg(TEST262_COMMIT)
```
Comment (Razican) [inline boa_tester/src/main.rs:268]: We could just default to the latest commit in `main`, and provide a commit from the CLI. We could then hard-code this in our CI if needed.

Comment (jasonwilliams) [reply boa_tester/src/main.rs:268]: agree, it seems odd to choose an arbitary commit. Why did we choose this one? Was it just the one we're using at the moment?

 I guess CI would have some environment variable to swap "main" out with whatever we're using.

**Comment by jasonwilliams** [comment]
Good idea, I don't see there being an issue with this. It should make the whole git submodules issue easier

File: boa_tester/src/main.rs (hunk 1 above)
Comment (gakonst) [reply boa_tester/src/main.rs:268]: Agreed - seems fine to default to "main" and override via CLI which sources the value from CI or other source the user controls.

**Review by gakonst — COMMENTED** [review]
Love it!! Thank you for taking this on, big devex upgrade for anyone using boa as a dep and doesn't want their clone to take minutes ^^

**Comment by HalidOdat** [comment]
This is now complete, and ready for review/merge :)

Now tester warns if local test262 clone is out of sync with `origin/main`, Added `--test262-update` for automatically updating the local repository and `--test262-commit` for specifying a commit.

**Comment by jedel1043** [comment]
Did you consider using [`git2`](https://docs.rs/git2/latest/git2/) instead of calling manual commands? Maybe that could further simplify the repo handling logic.

**Comment by gakonst** [comment]
Hi folks - wonder if we (we = [Reth](https://github.com/paradigmxyz/reth/) dev team) can help with this PR in any way?

**Review by jedel1043 — CHANGES_REQUESTED** [review]
This looks great implementation-wise, but I'd argue we should have a way to pin the commit downloaded by the tool for reproducibility in our CI. Maybe a file that lives in the root of the repo?

**Comment by HalidOdat** [comment]
> This looks great implementation-wise, but I'd argue we should have a way to pin the commit downloaded by the tool for reproducibility in our CI. Maybe a file that lives in the root of the repo?

Maybe merge `test_ignore.toml` file into something like `test262_config.toml`, to avoid having more than one config file `boa_tester`

**Comment by jedel1043** [comment]
> Maybe merge `test_ignore.toml` file into something like `test262_config.toml`, to avoid having more than one config file `boa_tester`

Great idea! We just have to create a new `ignore` section and we can put a `commit` property on the root

**Comment by jasonwilliams** [comment]
Hey @gakonst what does Reth use Boa for? I’m curious

**Comment by HalidOdat** [comment]
> > Maybe merge `test_ignore.toml` file into something like `test262_config.toml`, to avoid having more than one config file `boa_tester`
> 
> Great idea! We just have to create a new `ignore` section and we can put a `commit` property on the root

Implemented in the latest commit https://github.com/boa-dev/boa/pull/3214/commits/3a0d75cb799f96bb0153995569d0ffdedf569316 :)

**Comment by gakonst** [comment]
> Hey @gakonst what does Reth use Boa for? I’m curious

Hi Jason, Reth (w/o getting in too many of the details around the whole thing) exposes a debugging JSON-RPC service to provide "execution traces" of the request, showing step by step what happened. These traces can be configured to run with certain "tracers", some of which are part of the node. However, there's benefit in allowing users to specify custom tracers. The way this was implemented originally in go-ethereum was by embedding a Javascript sandbox and on various parts of the tracer lifecycle to call to some client-side provided JS that augments the tracer output.

In Reth, we use boa to provide that JS sandbox to the tracer. See [Go-ethereum's docs](https://geth.ethereum.org/docs/developers/evm-tracing/built-in-tracers#Javascript%20tracers) for more on this.

Hope that's helpful, happy to provide other context if needed but do not want to derail this issue.

File: boa_tester/src/main.rs [hunk 2]
```diff
@@ -204,35 +232,58 @@ enum Cli {
...
+        } => {
+            let config: Config = {
+                let input = std::fs::read_to_string(config_path)?;
+                toml::from_str(&input).wrap_err("could not decode tester config file")?
+            };
+
+            let test262_commit = match (test262_commit.as_deref(), config.commit()) {
+                (Some("latest"), _) | (None, "" | "latest") => None,
+                (Some(commit), _) | (None, commit) => Some(commit),
+            };
```
Comment (jedel1043) [inline boa_tester/src/main.rs:261]:
```suggestion
            let test262_commit = test262_commit
                .as_deref()
                .or_else(|| Some(config.commit()))
                .filter(|s| !["", "latest"].contains(s));
```

**Review by jedel1043 — APPROVED** [review]
Nice work! I just have a small nitpick that doesn't block merging this :)

**Review by nekevss — APPROVED** [review]
looks good to me!

**Pull request merged**